src/EventStore.Core/Caching/DynamicCacheManager.cs

```diff
@@ -0,0 +1,200 @@
+					cacheSettings.InitialMaxMemAllocation = allocatedMem;
+					Log.Information(
+						"{name} cache size auto-configured to ~{allocatedMem:N0} bytes.",
+						cacheSettings.Name, allocatedMem);
+					_maxMemAllocation[cacheIndex] = allocatedMem;
+				}
+			}
+		}
+
+		private void ResizeCachesIfNeeded() {
+			var freeMem = _getFreeMem();
+			var keepFreeMem = Math.Max(_keepFreeMemBytes, _totalMem * _keepFreeMemPercent / 100);
+
+			if (freeMem >= keepFreeMem && DateTime.UtcNow - _lastResize < _minResizeInterval)
+				return;
+
+			if (freeMem < keepFreeMem) {
+				Log.Debug("Available system memory is lower than "
+				          + "{thresholdPercent}% or {thresholdBytes:N0} bytes: {freeMem:N0} bytes. Resizing caches.",
+					_keepFreeMemPercent, _keepFreeMemBytes, freeMem);
+			}
+
+			try {
+				var cachedMem = _cachesSettings.Sum(cacheSettings => cacheSettings.GetMemoryUsage());
+				ResizeCaches(freeMem, cachedMem);
+			} finally {
+				_lastResize = DateTime.UtcNow;
+			}
+		}
+
+		private void ResizeCaches(long freeMem, long cachedMem) {
+			var availableMem = CalcAvailableMemory(freeMem, cachedMem);
+
+			var sw = new Stopwatch();
+			var cacheIndex = -1;
+
+			foreach (var cacheSettings in _cachesSettings) {
+				cacheIndex++;
+
+				if (!cacheSettings.IsDynamic)
+					continue;
+
+				var allocatedMem = CalcMemAllocation(availableMem, cacheSettings.Weight, cacheSettings.MinMemAllocation);
+
+				// do not resize if the amount of memory allocated to the cache hasn't changed
+				if (_maxMemAllocation[cacheIndex] == allocatedMem)
+					continue;
+
+				sw.Restart();
+				cacheSettings.UpdateMaxMemoryAllocation(allocatedMem);
+				sw.Stop();
+				Log.Debug(
+					"{name} cache resized to ~{allocatedMem:N0} bytes in {elapsed}.",
+					cacheSettings.Name, allocatedMem, sw.Elapsed);
+				_maxMemAllocation[cacheIndex] = allocatedMem;
+			}
+		}
+
+		private long CalcAvailableMemory(long freeMem, long cachedMem) {
+			var keepFreeMem = Math.Max(_keepFreeMemBytes, _totalMem * _keepFreeMemPercent / 100);
+			var availableMem = Math.Max(0L, freeMem + cachedMem - keepFreeMem);
+
+			Log.Debug($"Calculating memory available for caching based on:\n" +
+			          "Free memory: {freeMem:N0} bytes\n" +
+			          "Total memory: {totalMem:N0} bytes\n" +
+			          "Cached memory: ~{cachedMem:N0} bytes\n" +
+			          "Keep free %: {keepFreeMemPercent}%\n" +
+			          "Keep free bytes: {keepFreeMemBytes:N0} bytes\n\n" +
+			          "Memory available for caching: ~{availableMem:N0} bytes\n",
+				freeMem, _totalMem, cachedMem,
+				_keepFreeMemPercent, _keepFreeMemBytes, availableMem);
+
+			return availableMem;
+		}
+
+		private long CalcMemAllocation(long availableMem, int cacheWeight, long minMemAllocation) {
+			return Math.Max(availableMem * cacheWeight / _totalWeight, minMemAllocation);
+		}
+
+		private void Tick() {
+			_bus.Publish(
+				TimerMessage.Schedule.Create(
```

Comment:
the schudule message could be a member and reused each time (same as in InaugurationManager)

Reply:
fixed in: 58281fc4812c1c8c80ada6f0ddc0c3716254f7a5